analyzer/src/main/kotlin/managers/utils/DependencyGraphBuilder.kt

```diff
@@ -0,0 +1,324 @@
+ * correctly.
+ *
+ * This class implements the full logic to construct a [DependencyGraph], independent on the concrete representation of
+ * dependencies used by specific package managers. To make this class compatible with such a dependency representation,
+ * the package manager implementation has to provide a [DependencyHandler]. Via this handler, all the relevant
+ * information about dependencies can be extracted.
+ */
+class DependencyGraphBuilder<D>(
+    /**
+     * The [DependencyHandler] used by this builder instance to extract information from the dependency objects when
+     * constructing the [DependencyGraph].
+     */
+    val dependencyHandler: DependencyHandler<D>
+) {
+    companion object {
+        /**
+         * Internal class to represent the result of a search in the dependency graph. The outcome of the search
+         * determines how to integrate a specific dependency into the dependency graph.
+         */
+        private sealed class GraphSearchResult {
+
+            /**
+             * A specialized [GraphSearchResult] that indicates that the dependency that was searched for is already
+             * present in the dependency graph. This is the easiest case, as the [DependencyReference] that was found
+             * can directly be reused.
+             */
+            data class Found(
+                /** The reference to the dependency that was searched in the graph. */
+                val ref: DependencyReference
+            ) : GraphSearchResult()
+
+            /**
+             * A specialized [GraphSearchResult] that indicates that the dependency that was searched for was not found
+             * in the dependency graph, but a fragment has been detected, to which it can be added.
+             */
+            data class NotFound(
+                /** The index of the fragment to which to add the dependency. */
+                val fragmentIndex: Int
+            ) : GraphSearchResult()
+
+            /**
+             * A specialized [GraphSearchResult] that indicates that the dependency that was searched for in its
+             * current form cannot be added to any of the existing fragments. This means that either there are no
+             * fragments yet or each fragment contains a variant of the dependency with an incompatible dependency
+             * tree. In this case, a new fragment has to be added to the graph to host this special variant of this
+             * dependency.
+             */
+            object Incompatible : GraphSearchResult()
+        }
+    }
+
+    /**
+     * A list storing the identifiers of all dependencies added to this builder. This list is then used to resolve
+     * dependencies based on their indices.
+     */
+    private val dependencyIds = mutableListOf<String>()
+
+    /**
+     * A map listing the dependencies known to this builder and their numeric indices. This is used for a fast
```

Comment:
Maybe: "A mapping of the identifiers of the dependencies known to this builder to their numeric indices."